Add a Dockerfile

piovere opened the pull request:
Added a docker file that builds and then serves the website. It worked on my arm Mac!

sjml commented on .gitignore:
```diff
@@ -7,3 +7,5 @@
 /scripts/deploy_frontend.sh
 /docs/notes/private.md
+example_bots_src/wanderer/.cache
+engine/rsc/lib/wasmbots_validator.wasm
```
No, the validator *should* be in the repo. At least that's what I'm rolling with for now. (Thinking of it as something akin to the beschi-generated files.)

piovere replied:
Roger that, put it back

sjml reviewed (changes requested):
made a few notes / questions

piovere commented on example_bots_src/wanderer/Makefile:
I had to do this because by default the devbox provided compiler was trying to put stuff into a nix directory to which it did not have access. I can update the script to clean up the cache directory when it's done.

sjml replied:
How is C stuff usually built with devbox? Does everyone just add a cache like this?

hardliner66 replied:
I think the bigger question here is if the project even wants to use devbox. Because if yes, it might not even be necessary to add docker, because you can use devbox to manage deps, etc.

sjml replied:
I have no strong opinions on the matter one way or the other... I know Docker has been The Way™ for a long while, but I'm not especially tied to it. From what I understand devbox is a nicer experience, but maybe less familiar? 

I'm in favor of whatever will help more people get things running faster. :)

hardliner66 replied:
@sjml I don't mind either, but there should be a clear understanding on what is added and why. And there should be a "canonical" way to build and run the project that people can use.

As far as I see it, docker has the benefit that most people will already be familiar with it, so it should be provided either way. For devbox: it seems really cool, so if you think its nice to use locally to set up dev environments, then its definitely a way to go.

sjml replied:
Agreed on both counts -- a canonical setup process is definitely a Good Thing, and Docker is, for good and ill, the standard way of such things. Until now I've just been using my own ad-hoc local dev environment, which of course "works on my machine" but not great for collaboration. 

I'm going to put this in a discussion since I think we're already stretching the format of comments on a PR here... will link it once I do.

sjml replied:
Discussion here: https://github.com/sjml/wasmbots/discussions/7

hardliner66 commented:
```
> docker run wasmbots
info: using existing install for 'stable-x86_64-unknown-linux-gnu'
info: default toolchain set to 'stable-x86_64-unknown-linux-gnu'

  stable-x86_64-unknown-linux-gnu unchanged - rustc 1.84.1 (e71f9a9a9 2025-01-27)

info: component 'rust-std' for target 'wasm32-unknown-unknown' is up to date
/code/.devbox/gen/scripts/serve.sh: 7: cd: can't cd to validator
/bin/bash: scripts/_build_wasms.sh: No such file or directory
Error: error running script "serve" in Devbox: exit status 127
```

Also, I would recommend to create a new branch next time you do a PR. Otherwise your main will be dirty and you cannot work on other things while you wait for a PR to get merged :)

sjml commented:
Let's hold off on this for a bit until we hash out the goals in this discussion. https://github.com/sjml/wasmbots/discussions/7

piovere commented:
'docker run wasmbots' won't work as is--you have to mount the code volume. I can set it up to copy into the container though.

hardliner66 commented:
> 'docker run wasmbots' won't work as is--you have to mount the code volume. I can set it up to copy into the container though.

I quickly put together a Dockerfile without devbox. Just to have a reference as well. https://github.com/hardliner66/wasmbots/blob/add_docker/Dockerfile

I think it would be nice to have a container that people can just run, so copying in the necessary stuff is a good idea.

piovere commented:
I went with devbox because I was having trouble with docker running x86 containers on my machine so I hoped adding a dependency manager would abstract that away a bit :/

piovere commented:
updated based on discussion in #7. Dockerfile now copies the files, builds them, then passes them off to a caddy image that serves them on http.

Also cleaned the .gitignore and cache instructions commented on above

sjml commented:
> I quickly put together a Dockerfile without devbox. Just to have a reference as well. https://github.com/hardliner66/wasmbots/blob/add_docker/Dockerfile

This is great to see, and especially nice to have the reference of getting everything up from scratch. Maybe will end up with multiple Dockerfiles both for the approaches and, as we talked about in the discussion, for different use cases. 

Curious: were you having problems that were resolved by `svelte-kit sync`? I just noted you added that with the `npm run prepare` step. I didn't think that would be necessary since it gets done automatically in the build step.

sjml commented:
Made some changes that runs ok locally for me -- check it out with Docker/devbox on your end.

sjml reviewed (approved):
LGTM

hardliner66 commented:
> > I quickly put together a Dockerfile without devbox. Just to have a reference as well. https://github.com/hardliner66/wasmbots/blob/add_docker/Dockerfile
> 
> This is great to see, and especially nice to have the reference of getting everything up from scratch. Maybe will end up with multiple Dockerfiles both for the approaches and, as we talked about in the discussion, for different use cases.
> 
> Curious: were you having problems that were resolved by `svelte-kit sync`? I just noted you added that with the `npm run prepare` step. I didn't think that would be necessary since it gets done automatically in the build step.

Sorry for the late reply.

The tsconfig.json references a file from svelte-kit `"extends": "./.svelte-kit/tsconfig.json",` which wasn't created automatically inside my docker build, so I added the prepare step for svelte-kit to create the needed files.

sjml commented:
> The tsconfig.json references a file from svelte-kit `"extends": "./.svelte-kit/tsconfig.json",` which wasn't created automatically inside my docker build, so I added the prepare step for svelte-kit to create the needed files.

Ah, ok. You must have recreated the `package-lock.json` file at some point because it should have been pinned to an earlier version where that wasn't necessary yet. (By happy coincidence, I was updating dependencies tonight and ran into this same issue when updating svelte-kit, so already knew where to look. So belated thanks! :D)
